mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -49,152 +49,151 @@ def _autolog(log_every_n_epoch=1, log_models=True):
+                    if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                        self.early_stopping = True
+                        self._log_early_stop_params(callback)
+
+                # TODO For logging optimizer params - Following scenarios are to revisited.
+                # 1. In the current scenario, only the first optimizer details are logged.
+                #    Code to be enhanced to log params when multiple optimizers are used.
+                # 2. mlflow.log_params is used to store optimizer default values into mlflow.
+                #    The keys in default dictionary are too short, Ex: (lr - learning_rate).
+                #    Efficient mapping technique needs to be introduced
+                #    to rename the optimizer parameters based on keys in default dictionary.
+
+                if hasattr(trainer, "optimizers"):
+                    optimizer = trainer.optimizers[0]
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+
+                    if hasattr(optimizer, "defaults"):
+                        try_mlflow_log(mlflow.log_params, optimizer.defaults)
+
+                summary = str(ModelSummary(pl_module, mode="full"))
+                tempdir = tempfile.mkdtemp()
+                try:
+                    summary_file = os.path.join(tempdir, "model_summary.txt")
+                    with open(summary_file, "w") as f:
+                        f.write(summary)
+
+                    try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
+                finally:
+                    shutil.rmtree(tempdir)
+
+            def on_train_end(self, trainer, pl_module):
+                """
+                Logs the model checkpoint into mlflow - models folder on the training end
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                if log_models:
+                    mlflow.pytorch.log_model(pytorch_model=trainer.model, artifact_path="model")
+
+                    if self.early_stopping and trainer.checkpoint_callback.best_model_path:
+                        try_mlflow_log(
+                            mlflow.log_artifact,
+                            local_path=trainer.checkpoint_callback.best_model_path,
+                            artifact_path="restored_model_checkpoint",
+                        )
+
+            def on_test_end(self, trainer, pl_module):
+                """
+                Logs accuracy and other relevant metrics on the testing end
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                try_mlflow_log(mlflow.set_tag, "Mode", "testing")
+                for key, value in trainer.callback_metrics.items():
+                    try_mlflow_log(mlflow.log_metric, key, float(value))
```

Comment:
Should we use `record_metrics` here? (Probably not too significant since this is presumably called at the end of training)

Reply:
No, because after extensive  reading of pytorch docs, from my understanding: testing stage which invokes `on_test_end ` is not part of the training stage whether it be `fit` or `train`. Therefore the batch metrics logger context would have been torn down, and recording metrics using it would not make sense. I'm sure you would know better, if I can guarantee that `on_test_end()` will be called before the context is teared down then definitely we should use `record_metrics()` instead of `log_metric `.